### PR TITLE
Implement iterator over nodes; refactor key-value iterator

### PR DESCRIPTION
**Pull request by danlaine**
alternate implementation to #499.

This PR implements an iterator over nodes. The iterator can be configured to start at a given key. Uses this node iterator to refactor the key-value iterator implementation.

### CODE REVIEW

File: firewood/src/merkle/stream.rs
Comment (richardpringle) [inline firewood/src/merkle/stream.rs:200]: Can you use `Vec::with_capacity` here to minimize reallocations?

Comment (richardpringle) [reply firewood/src/merkle/stream.rs:200]: 🤔... maybe that wouldn't be good

File: firewood/src/merkle/stream.rs
Comment (richardpringle) [inline firewood/src/merkle/stream.rs:196]: Can you use `node.as_ptr()` instead? Or is there a time when `node_addr` does not point to `node`?

File: firewood/src/merkle/stream.rs
Comment (richardpringle) [inline firewood/src/merkle/stream.rs:267]: nit: 
`is_prefix` implies a `bool` as a return type. I say this as someone who frequently gives functions terrible names. What about `compare_as_prefix` or something along those lines?

File: firewood/src/merkle/stream.rs
Comment (richardpringle) [inline firewood/src/merkle/stream.rs:285]: 🤔 , did `child` or `child_addr` change here? Don't you already have the `child`?

File: firewood/src/merkle/stream.rs
Comment (richardpringle) [inline firewood/src/merkle/stream.rs:293]: nit: I think it's easier to read single-line property assigns
```suggestion
                        let key = matched_key_nibbles
                            .iter()
                            .copied()
                            .chain(partial_key.iter().copied())
                            .collect();
                        iter_stack.push(IterationNode::Unvisited {
                            key,
                            node: child,
                        });
```

File: firewood/src/merkle/stream.rs
Comment (richardpringle) [inline firewood/src/merkle/stream.rs:318]: I'm surprised clippy didn't catch this, but it's just an if-statement

```suggestion
                if is_prefix(leaf.path.iter(), unmatched_key_nibbles).0 == Ordering::Greater {
                   let node = merkle
                        .get_node(node_addr)
                        .map_err(|e| api::Error::InternalError(Box::new(e)))?;

                    // The leaf's key > `key`, so we can stop here.
                    let key = matched_key_nibbles
                        .iter()
                        .copied()
                        .chain(leaf.path.iter().copied())
                        .collect();
                    
                    iter_stack.push(IterationNode::Unvisited {
                        key,
                        node,
                    });
                }
```

File: firewood/src/merkle/stream.rs
Comment (richardpringle) [inline firewood/src/merkle/stream.rs:307]: 🤔, am I missing something or are you redundantly populating `node`?

File: firewood/src/merkle/stream.rs
Comment (richardpringle) [inline firewood/src/merkle/stream.rs:25]: I think you can also use triple `/` here too
```suggestion
    /// This node has not been returned yet.
    Unvisited {
        /// The key (as nibbles) of this node.
        key: Key,
        node: NodeObjRef<'a>,
    },
    /// This node has been returned. Track which child to visit next.
```

File: firewood/src/merkle/stream.rs
Comment (richardpringle) [inline firewood/src/merkle/stream.rs:160]: 🥇

File: firewood/src/merkle/stream.rs
Comment (richardpringle) [inline firewood/src/merkle/stream.rs:167]: 🤔 Already have the child again?

**Review by richardpringle — CHANGES_REQUESTED** [review]
So, I really don't want to send you chasing any red herrings, but *I think* you can minimize the usage of `node_addr`/`child_addr` such that you aren't unnecessarily fetching nodes.

File: firewood/src/merkle/stream.rs
Comment (danlaine) [reply firewood/src/merkle/stream.rs]: removed

File: firewood/src/merkle/stream.rs
Comment (danlaine) [reply firewood/src/merkle/stream.rs]: renamed `compare_partial_path`

File: firewood/src/merkle/stream.rs
Comment (danlaine) [reply firewood/src/merkle/stream.rs]: removed

File: firewood/src/merkle/stream.rs
Comment (danlaine) [reply firewood/src/merkle/stream.rs]: removed

File: firewood/src/merkle/stream.rs
Comment (danlaine) [reply firewood/src/merkle/stream.rs]: removed `node_addr` -- not needed

File: firewood/src/merkle/stream.rs
Comment (danlaine) [reply firewood/src/merkle/stream.rs:195]: 🤷 Honestly you would know better than me lol

File: firewood/src/merkle/stream.rs
Comment (danlaine) [reply firewood/src/merkle/stream.rs]: I'll defer to you and the rest of the team on how to handle this. For the time being I removed the TODO since I don't think it's a blocker.

File: firewood/src/merkle/stream.rs
Comment (richardpringle) [reply firewood/src/merkle/stream.rs:195]: I was thinking that `with_capacity` would be better if you know the max size, but you're cloning the `matched_key_nibbles` as it grows, so you don't want a bunch of oversized vectors that will never grow. That would cause excess memory usage in an environment where memory is prized.

File: firewood/src/merkle/stream.rs
Comment (rkuris) [inline firewood/src/merkle/stream.rs:244]:
```suggestion
                let Some(child_addr) = branch.children[nib as usize] else {
                    // There is no child at `nib`.
                    // We'll visit `node`'s first child at index > `nib`
                    // first (if it exists).
                    return Ok(MerkleNodeStreamState::Initialized { iter_stack });
                };
```

**Review by rkuris — CHANGES_REQUESTED** [review]

File: firewood/src/merkle/stream.rs
Comment (richardpringle) [inline firewood/src/merkle/stream.rs:228]: For a follow up PR:
I think it might make sense to do the following:
```suggestion
                    key: matched_key_nibbles.iter().copied().collect(),
```
I think you did this elsewhere in the PR too. The reason that I would suggest to do this instead of calling `clone` for turning a `Vec` into a `Box<[u8]>`, is that there are no guarantees that `Vec::clone` doesn't also create a new `Vec` with the same capacity, then turn it into a `Box<[u8]>`, dropping the extra capacity. 

In practice, that's not how the current version of `Vec::clone` works, but the only thing that's guaranteed is the API between Rust versions. 

I'm sure that you get the exact same output assembly, but I think it's generally better to be explicit about size (by using an iterator).

File: firewood/src/merkle/stream.rs
Comment (richardpringle) [inline firewood/src/merkle/stream.rs:567]: I strongly suggest you split these test cases up so that they are run as separate tests instead of checking different cases all in the same test. You can always have a separate function to create the same trie, but having a failing test case prevent a subsequent test does not lead to a great debugging experience.

**Review by richardpringle — APPROVED** [review]

File: firewood/src/merkle.rs
Comment (danlaine) [reply firewood/src/merkle.rs:1732]: I thought it made sense to expose this new functionality even though it's not used right now, but I can remove it since it's not used right now

Comment (danlaine) [reply firewood/src/merkle.rs:1732]: Hmm if I remove `#[allow(dead_code)]` then it complains that the code is unused, but if I remove the function, I can't test the node iterator, which I should do because it's used by the key-value iterator

File: firewood/src/merkle/stream.rs
Comment (danlaine) [reply firewood/src/merkle/stream.rs]: agreed, kind of redundant. Removed.

File: firewood/src/merkle/stream.rs
Comment (rkuris) [reply firewood/src/merkle/stream.rs]: Another good choice is: `key: Box::from(matched_key_nibbles.as_slice()),` since [Box::from can take a slice](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.from-20)

File: firewood/src/merkle.rs
Comment (rkuris) [reply firewood/src/merkle.rs:1732]: Mark them #[cfg(test)] if they are only used in testing, or perhaps only have them defined in the test module.

Comment (danlaine) [reply firewood/src/merkle.rs]: moved to test module and did a bit of renaming

File: firewood/src/merkle/stream.rs
Comment (danlaine) [reply firewood/src/merkle/stream.rs]: added

File: firewood/src/merkle/stream.rs
Comment (rkuris) [inline firewood/src/merkle/stream.rs:262]: nit; doubtful it makes any difference but seems easier to digest
```suggestion
                        let key = matched_key_nibbles
                            .iter()
                            .chain(partial_key.iter())
                            .copied()
                            .collect();
```

File: firewood/src/merkle/stream.rs
Comment (rkuris) [inline firewood/src/merkle/stream.rs:274]: nit: since we're doing exactly the same thing earlier, perhaps a helper function would be useful here

File: firewood/src/merkle/stream.rs
Comment (rkuris) [inline firewood/src/merkle/stream.rs:430]: nit: this seems a little odd, in that we had to reverse the output of `enumerate()`. Seems like these tuples should be `(u8, DiskAddress)` to match the way `enumerate()` works.

nit: naming could be better. Typically `get` is not used in rust. Some ideas:
- `enumerated_children_iter`
- `as_enumerated_children_iter`

File: firewood/src/merkle/stream.rs
Comment (rkuris) [inline firewood/src/merkle/stream.rs:749]: nit: new test?

**Review by rkuris — APPROVED** [review]
Approved, all are remaining nits. There is one extra allocation you should probably remove before merging.

**Review by richardpringle — APPROVED** [review]

**Review by richardpringle — APPROVED** [review]

**Comment by richardpringle** [comment]
🥇

**Pull request merged**
